From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: JellySquid <jellysquid+atwork@protonmail.com>
Date: Tue, 7 Dec 2021 19:22:09 -0500
Subject: [PATCH] lithium MixinDirection

Patch from Yatopia
Original code by JellySquid, licensed under GNU Lesser General Public License v3.0
you can find the original code on https://github.com/CaffeineMC/lithium-fabric/ (Yarn mappings)

Updated to 1.18 by: Steven <steven@nevets.tech>

diff --git a/src/main/java/net/minecraft/core/Direction.java b/src/main/java/net/minecraft/core/Direction.java
index 8aeec2132138bf4b19a9042340c8784df4afda65..f7189180f2095704e754120a5e8471224bf96801 100644
--- a/src/main/java/net/minecraft/core/Direction.java
+++ b/src/main/java/net/minecraft/core/Direction.java
@@ -190,8 +190,12 @@ public enum Direction implements StringRepresentable {
         return var10000;
     }
 
+    /**
+     * @reason Avoid the modulo/abs operations
+     * @author JellySquid
+     */
     public Direction getOpposite() {
-        return from3DDataValue(this.oppositeIndex);
+        return VALUES[this.oppositeIndex];
     }
 
     public Direction getClockWise(Direction.Axis axis) {
@@ -452,8 +456,13 @@ public enum Direction implements StringRepresentable {
         return (float)((this.data2d & 3) * 90);
     }
 
+    /**
+     *
+     * @reason Do not allocate an excessive number of Direction arrays
+     * @author JellySquid
+     */
     public static Direction getRandom(Random random) {
-        return Util.getRandom(VALUES, random);
+        return VALUES[random.nextInt(VALUES.length)];
     }
 
     public static Direction getNearest(double x, double y, double z) {
